### PR TITLE
Safari Extension Handler implementation

### PR DESCRIPTION
**Pull request by JAHMCR**
Implemented an initial `SafariExtensionHandler` for communication between the browser and the Safari extension.

This PR will enable support for HTTP requests, while transfer monitoring will be implemented in the next PR.

### CODE REVIEW

File: src/helpers/safari-extension.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
+import { v4 as uuidv4 } from 'uuid';
```
Comment (dwosk) [inline src/helpers/safari-extension.ts:1]: We have `randomUuid` in helpers that you can use so we don't need to add a hard dependency.

File: src/helpers/helpers.ts [hunk 2]
```diff
@@ -177,6 +177,19 @@ export const isValidURL = (url: string): boolean => {
...
+ * @returns {boolean} Whether the browser is Safari.
+ */
+export const isSafari = () => {
+  const ua = navigator.userAgent.toLowerCase();
+  if (ua.indexOf('safari') !== -1) {
+    return ua.indexOf('chrome') === -1;
+  } else {
+    return false;
+  }
+};
```
Comment (dwosk) [inline src/helpers/helpers.ts:191]: We can use the same as that comes from Starter.

```suggestion
export const isSafari = (): boolean => {
  // eslint-disable-next-line
  return /^((?!chrome|android).)*safari/i.test(navigator.userAgent) && !(window as any).MSStream;
};
```

File: src/app/core.ts [hunk 3]
```diff
@@ -1,9 +1,47 @@
...
+  safariExtensionHandler = new SafariExtensionHandler();
+}
+
+function dispatchRequest(method: string, payload?: any): Promise<any> {
+  if (safariExtensionHandler) {
+    return safariExtensionHandler.sendRPCRequest(method, payload);
+  }
+
+  return client.request(method, payload);
+}
```
Comment (dwosk) [inline src/app/core.ts:44]: Could we have the http client and safari client implement the same interface? Then we can keep `client.request` throughout and just rely on the conditional above to assign the right one at runtime.

It seems like both the safari and http clients should just expose a `request` method of some sort that returns a Promise.

File: package.json [hunk 4]
```diff
@@ -45,6 +46,7 @@
     "webpack-dev-server": "^4.11.1"
   },
   "dependencies": {
-    "json-rpc-2.0": "^1.6.0"
+    "json-rpc-2.0": "^1.6.0",
+    "uuid": "^10.0.0"
```
Comment (dwosk) [inline package.json:50]: Can remove `uuid` and `@types/uuid` and just use the `randomUUID` from helpers.

File: src/helpers/client.ts [hunk 5]
```diff
@@ -77,9 +77,4 @@ class Client {
   };
 }
 
-export const client = new Client();
-
-export default {
-  client,
-  handlePromiseLikeErrors,
-};
+export default Client;
```
Comment (dwosk) [inline src/helpers/client.ts:80]: I think it's OK to still export `Client` singleton still. 

```suggestion
export const client = new Client();

export default {
  client,
  handlePromiseLikeErrors,
};
```

File: src/helpers/safari-extension.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+   * and delegates handling to the handleSafariExtensionEvent method.
+   */
+  private listenSafariExtensionEvents() {
+    document.addEventListener('AsperaDesktop.Response', (event: CustomEvent<JSONRPCResponse>) => {
+      this.handleSafariExtensionEvent(event.detail);
+    });
+  }
+}
+
+export default SafariExtensionHandler;
```
Comment (dwosk) [inline src/helpers/safari-extension.ts:135]: Export a singleton instance of `SafariExtensionHandler` as well?

Minor nitpick but for consistency (since we have `Client` already) should we rename this to `SafariClient`? Or have `HttpClient` and `SafariClient`?

File: tests/client.spec.ts [hunk 7]
```diff
@@ -1,7 +1,8 @@
-import {client} from '../src/helpers/client';
-import {asperaDesktop} from '../src/index';
 import {mockFetch} from './mocks';
+import {asperaDesktop} from '../src';
+import Client from '../src/helpers/client';
 
+const client = new Client();
```
Comment (dwosk) [inline tests/client.spec.ts:5]: If we still export a singleton you can revert these changes.

**Review by dwosk — CHANGES_REQUESTED** [review]

File: src/helpers/helpers.ts (hunk 2 above)
Comment (JAHMCR) [reply src/helpers/helpers.ts:191]: Done!

File: src/helpers/safari-extension.ts (hunk 1 above)
Comment (JAHMCR) [reply src/helpers/safari-extension.ts:1]: Done!

File: package.json (hunk 4 above)
Comment (JAHMCR) [reply package.json:50]: Done!

File: src/helpers/client.ts (hunk 5 above)
Comment (JAHMCR) [reply src/helpers/client.ts:80]: Done!

File: src/helpers/safari-extension.ts (hunk 6 above)
Comment (JAHMCR) [reply src/helpers/safari-extension.ts:135]: Done!

File: tests/client.spec.ts (hunk 7 above)
Comment (JAHMCR) [reply tests/client.spec.ts:5]: Done!

File: src/app/core.ts [hunk 8]
```diff
@@ -1,9 +1,32 @@
...
+  generateErrorBody,
+  generatePromiseObjects,
+  getWebsocketUrl,
+  isSafari,
+  isValidTransferSpec,
+  randomUUID,
+  throwError
+} from '../helpers/helpers';
+import {httpClient} from '../helpers/client/http-client';
+import {safariClient} from '../helpers/client/safari-client';
```
Comment (dwosk) [inline src/app/core.ts:14]: You could re-export both of these from `client` module to clean this up a bit. Then it would be:
```suggestion
import {httpClient, safariClient} from '../helpers/client';
```

File: src/helpers/client/http-client.ts [hunk 9]
```diff
@@ -64,22 +64,22 @@ class HttpClient {
...
+    return handlePromiseLikeErrors(this.httpClient.request(method, payload));
   };
 }
 
-export const client = new Client();
+export const httpClient = new HttpClient();
 
 export default {
-  client,
+  client: httpClient,
```
Comment (dwosk) [inline src/helpers/client/http-client.ts:83]:
```suggestion
  httpClient,
```
I think?

File: tests/client.spec.ts [hunk 10]
```diff
@@ -1,6 +1,10 @@
-import {client} from '../src/helpers/client';
-import {asperaDesktop} from '../src/index';
 import {mockFetch} from './mocks';
+import {asperaDesktop} from '../src';
+import {isSafari} from '../src/helpers/helpers';
+import {httpClient} from '../src/helpers/client/http-client';
+import {safariClient} from '../src/helpers/client/safari-client';
+
+const client: Client = isSafari() ? safariClient : httpClient;
```
Comment (dwosk) [inline tests/client.spec.ts:7]: I'm not sure checking for Safari has any benefit in the specs. Would this ever return true?

File: package-lock.json [hunk 11]
```diff
@@ -8039,6 +8039,15 @@
...
     },
+    "node_modules/sockjs/node_modules/uuid": {
+      "version": "8.3.2",
+      "resolved": "https://registry.npmjs.org/uuid/-/uuid-8.3.2.tgz",
+      "integrity": "sha512-+NYs2QeMWy+GWFOEm9xnn6HCDp0l7QBD7ml8zLUmJ+93Q5NF0NocErnwkTkXVFNiX3/fpC6afS8Dhb/gz7R7eg==",
+      "dev": true,
+      "bin": {
+        "uuid": "dist/bin/uuid"
+      }
+    },
```
Comment (dwosk) [inline package-lock.json:8050]: Make sure you re-run `npm install` or remove `uuid` via `npm uninstall` to remove the dependency from the lockfile too.

**Review by dwosk — CHANGES_REQUESTED** [review]

File: src/helpers/client/http-client.ts (hunk 9 above)
Comment (JAHMCR) [reply src/helpers/client/http-client.ts:83]: Done!

File: tests/client.spec.ts (hunk 10 above)
Comment (JAHMCR) [reply tests/client.spec.ts:7]: Done!

File: package-lock.json (hunk 11 above)
Comment (JAHMCR) [reply package-lock.json:8050]: Done!

Comment (JAHMCR) [reply package-lock.json:8050]: This was already there, it's part of another dependency

File: src/app/core.ts (hunk 8 above)
Comment (JAHMCR) [reply src/app/core.ts]: Done!

File: package-lock.json (hunk 11 above)
Comment (dwosk) [reply package-lock.json:8050]: If we're not updating packge.json file then there shouldn't be any changes to the lockfile. So it seems like the addition of uuid in our stuff was not properly taken out. I think you could just do a `git checkout package-lock.json` to revert it.

**Review by dwosk — APPROVED** [review]
